lib/plugin/opsgenie/client.go

```diff
@@ -0,0 +1,259 @@
+	client.SetHeader("Authorization", "GenieKey "+conf.APIKey)
+	client.SetHostURL(conf.APIEndpoint)
+	return &Client{
+		client:       client,
+		ClientConfig: conf,
+	}, nil
+}
+
+func errWrapper(statusCode int, err error) error {
+	switch statusCode {
+	case http.StatusForbidden:
+		return trace.AccessDenied("opsgenie API access denied: %v", err)
+	case http.StatusRequestTimeout:
+		return trace.ConnectionProblem(err, "connecting to opsgenie API")
+	}
+	return trace.Wrap(err)
+}
+
+// CreateAlert creates an opsgenie alert.
+func (og Client) CreateAlert(ctx context.Context, reqID string, reqData RequestData) (OpsgenieData, error) {
+	bodyDetails, err := buildAlertBody(og.WebProxyURL, reqID, reqData)
+	if err != nil {
+		return OpsgenieData{}, trace.Wrap(err)
+	}
+
+	body := AlertBody{
+		Message:     fmt.Sprintf("Access request from %s", reqData.User),
+		Alias:       fmt.Sprintf("%s/%s", alertKeyPrefix, reqID),
+		Description: bodyDetails,
+		Responders:  og.getResponders(reqData),
+		Priority:    og.Priority,
+	}
+
+	var result AlertResult
+	if resp, err := og.client.NewRequest().
+		SetContext(ctx).
+		SetBody(body).
+		SetResult(&result).
+		Post("alerts"); err != nil {
+		return OpsgenieData{}, errWrapper(resp.StatusCode(), err)
+	}
+
+	return OpsgenieData{
+		AlertID: result.Alert.ID,
+	}, nil
+}
+
+func (og Client) getResponders(reqData RequestData) []Responder {
+	schedules := og.DefaultSchedules
+	if reqSchedules, ok := reqData.RequestAnnotations[ReqAnnotationRespondersKey]; ok {
+		schedules = reqSchedules
+	}
+	responders := make([]Responder, 0, len(schedules))
+	for _, s := range schedules {
+		responders = append(responders, Responder{
+			Type: "schedule",
+			ID:   s,
+		})
+	}
+	return responders
+}
+
+// PostReviewNote posts a note once a new request review appears.
+func (og Client) PostReviewNote(ctx context.Context, alertID string, review types.AccessReview) error {
+	note, err := buildReviewNoteBody(review)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	body := AlertNote{
+		Note: note,
+	}
+	if resp, err := og.client.NewRequest().
+		SetContext(ctx).
+		SetBody(body).
+		SetPathParams(map[string]string{"alertID": alertID}).
+		SetQueryParams(map[string]string{"identifierType": "id"}).
+		Post("alerts/{alertID}/notes"); err != nil {
+		return errWrapper(resp.StatusCode(), err)
+	}
+	return nil
+}
+
+// ResolveAlert resolves an alert and posts a note with resolution details.
+func (og Client) ResolveAlert(ctx context.Context, alertID string, resolution Resolution) error {
+	note, err := buildResolutionNoteBody(resolution)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	body := AlertNote{
+		Note: note,
+	}
+	if resp, err := og.client.NewRequest().
+		SetContext(ctx).
+		SetBody(body).
+		SetPathParams(map[string]string{"alertID": alertID}).
+		SetQueryParams(map[string]string{"identifierType": "id"}).
+		Post("alerts/{alertID}/close"); err != nil {
+		return errWrapper(resp.StatusCode(), err)
```

Comment:
I don't think we want to use `resp` if the error is non-nil, [according to the docs](https://pkg.go.dev/net/http#Client.Do): 

> On error, any Response can be ignored. A non-nil Response with a non-nil error only occurs when CheckRedirect fails, and even then the returned Response.Body is already closed. 

Also, non-2XX status code will not cause an error, I think we want to check if the call succeeded.

> An error is returned if caused by client policy (such as CheckRedirect), or failure to speak HTTP (such as a network connectivity problem). A non-2xx status code doesn't cause an error. 

Can you add a test case where the HTTP call succeeds but the server returns an error?